### ISSUE TITLE
RBAC on AKS: Helm keeps asking for login token on every command

### ISSUE THREAD

**Issue by dploeger** [helm/helm#5788]
We're on AKS with RBAC enabled. Now when running any helm command that connects to the k8s cluster, I have to reauthenticate. This is the output:

```
To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code <redacted> to authenticate.
```

Then I can use helm for maybe one or two other commands and the the login message is displayed again.

After opening the page and authenticating, Helm continues without a problem, but it is very tedious like this.

This doesn't happen with kubectl.

Output of `helm version`:

```
Client: &version.Version{SemVer:"v2.14.0", GitCommit:"05811b84a3f93603dd6c2fcfe57944dfa7ab7fd0", GitTreeState:"clean"}
To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code <redacted> to authenticate.
Server: &version.Version{SemVer:"v2.14.0", GitCommit:"05811b84a3f93603dd6c2fcfe57944dfa7ab7fd0", GitTreeState:"clean"}
```

Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.4", GitCommit:"c27b913fddd1a6c480c229191a087698aa92f0b1", GitTreeState:"clean", BuildDate:"2019-03-01T23:34:27Z", GoVersion:"go1.12", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.5", GitCommit:"2166946f41b36dea2c4626f90a77706f426cdea2", GitTreeState:"clean", BuildDate:"2019-03-25T15:19:22Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
AKS

**Comment by jdolitsky** [comment]
@dploeger I'm finding it hard to believe that Helm has anything to do with issue, api calls to Kubernetes are done in the same fashion as kubectl (same client libs and KUBECONFIG).

Would you mind opening this issue here instead? https://github.com/Azure/AKS

**Comment by bacongobbler** [comment]
I wouldn't be so quick to discard this as an upstream problem. :)

It's possible that there's a bug in helm when loading Azure credentials for use similar to kubectl. Do you see the same behaviour with kubectl, @dploeger?

**Comment by dploeger** [comment]
Yes, because kubectl works, I opened the issue here. Should I still double post it to the AKS-Repo as well?

I let helm run with debugging tomorrow and keep you posted if I find something there.

**Comment by jdolitsky** [comment]
@dploeger  sorry for jumping the gun :)

I think it may be helpful to ask there as well. They will probably have more background into what causes the the Azure-based kubeconfig to become invalid.

**Comment by dploeger** [comment]
Okay, will do then.

Hm. Is there any option to Helm that *really* enables debugging, because that is my debug output from `helm list`:

```
To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code <redacted> to authenticate.
[debug] Created tunnel using local port: '61806'

[debug] SERVER: "127.0.0.1:61806"

NAME         	REVISION	UPDATED                 	STATUS  	CHART               	APP VERSION	NAMESPACE
easykv-pim   	2       	Thu May 23 12:34:12 2019	DEPLOYED	easykv-pim-0.1.0    	1.0        	default
elasticsearch	1       	Thu May 23 12:22:58 2019	DEPLOYED	elasticsearch-1.27.2	6.7.0      	default
kibana       	1       	Thu May 23 12:23:08 2019	DEPLOYED	kibana-3.0.0        	6.7.0      	default
```

**Comment by dploeger** [comment]
I'm on macOS btw.

**Comment by dploeger** [comment]
Added Azure/AKS#1008

**Comment by dploeger** [comment]
Note: I'm using multiple accounts on Azure. Might be important.

**Comment by jorgechato** [comment]
Same thing here.
```bash
$ helm version

Client: &version.Version{SemVer:"v2.14.0", GitCommit:"05811b84a3f93603dd6c2fcfe57944dfa7ab7fd0", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.14.0", GitCommit:"05811b84a3f93603dd6c2fcfe57944dfa7ab7fd0", GitTreeState:"clean"}
```

```bash
$ kubectl version

Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.2", GitCommit:"17c77c7898218073f14c8d573582e8d2313dc740", GitTreeState:"clean", BuildDate:"2018-10-30T21:39:38Z", GoVersion:"go1.11.1", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.4", GitCommit:"c27b913fddd1a6c480c229191a087698aa92f0b1", GitTreeState:"clean", BuildDate:"2019-02-28T13:30:26Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
```

All ENV variables set properly and `~/.kube/config` in a good shape

**Comment by ErikLundJensen** [comment]
Try to upgrade az cli to the latest version (currently 2.0.65) and run
az login
az account set --subscription "MySubscription"
az aks get-credentials -n my-aks-cluster-name -g my-resource-group --overwrite-existing
kubectl get nodes

The first time you run kubectl you will get a challenge with link and token to MS device-login.
You should not get any challenges after that.

You may verify by running:
az logout
kubectl get nodes
(you should still see your nodes without getting a challenge)

This should also solve your Helm issues.

**Comment by dploeger** [comment]
It seems 2.0.64 [is the latest version](https://docs.microsoft.com/de-de/cli/azure/release-notes-azure-cli?view=azure-cli-latest) and that's the one I use and I used the commands you stated exactly.

Remember, that kubectl is working perfectly fine. Only Helm asks for a login constantly.

**Comment by ErikLundJensen** [comment]
Compare [latest version UK](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli?view=azure-cli-latest)
with the German version from your link.

**Comment by dploeger** [comment]
Interesting. However, the 2.0.65 isn't available on Homebrew currently 😞

**Comment by jorgechato** [comment]
@ErikLundJensen not sure how the az CLI could be related to this bug.
As @dploeger mentioned kubectl is working properly and if you downgrade helm to 2.7.2 for instance you don't have the annoying token request

**Comment by dploeger** [comment]
@jorgechato You don't? I have to try that. 😄

**Comment by jdolitsky** [comment]
In the Helm 2.14 [release notes](https://github.com/helm/helm/releases/tag/v2.14.0), I'm seeing `updated Kubernetes client libraries to v1.14` as a line item (`v1.14.1`), and it looks like you're running kubectl `1.13.4` @dploeger 

I suspect that if you upgrade/pin to kubectl to version `1.14.1` you might have the same problem with Azure auth?

If so our answer might be in what was introduced in kubernetes/kubernetes 1.14+

**Comment by dploeger** [comment]
@jdolitsky I'll check that out

**Comment by dploeger** [comment]
@jdolitsky I upgraded to kubectl 1.14.2 and got one token request. Until now, everything works fine. helm, too, actually. I keep you posted.

**Comment by jdolitsky** [comment]
bizarre.. I wonder if the token `az login` gives you needs to somehow be "upgraded" to be compatible w/ 2.14+

**Comment by jorgechato** [comment]
@jdolitsky no `az` CLI isn't related to this bug.
The problem looks like to be the combination of helm v.2.14 and kubectl v.2.13 (pending to test it in lower versions of kubectl).

Lower versions of helm (like v.2.7.2) and kubectl v.2.13 are working as expected.

**Comment by jorgechato** [comment]
I am just guessing but maybe if the k8s client used by helm is lower than the kubectl client, that  reproduce this bug

**Comment by jdolitsky** [comment]
@jorgechato you mean kubectl `v.1.13.x`? Sounds like `v.1.14.x` is indeed compatible w/ Helm 2.14

It might not be related to az CLI directly, but to Azure auth service. For example, Helm client attempts to auth against the cluster using a token created with 1.13 client (kubectl) and the auth layer responds with something like "please upgrade to use token for 2.14+"... I'm not sure.

**Comment by jorgechato** [comment]
my bad, yes the combination of helm v.2.14 and kubectl v.1.13.x

**Comment by jorgechato** [comment]
@jdolitsky I think what you described is what is going on there. Curious to know if the kubectl version is lower than the helm (k8s client) will trigger the same bug in kubectl

**Comment by bacongobbler** [comment]
According to a response in Slack in the #sig-azure channel, this appears to be an issue across the ecosystem. Here's the message:

> Hi folks. When using kubectl I'm forced to reauthenticate (To sign in, use a web browser to open the page https://microsoft.com/devicelogin) each time right after I used another cli tool (for example kubeseal) that uses the go-client library.

And here's another user's response:

> I started seeing this behavior recently too, once I upgraded my version of kubectl to 1.14. Any other CLI tools (kubeseal, helm, fluxctl, etc) that use an older(?) version of client-go seem to invalidate the login token and force me to go through the oauth device code flow again.

So it appears on first glance that the kube config format was changed between client-go 1.13 and client-go 1.14, introducing a backwards-incompatible change.

As others have mentioned before, it's probably best to stick to Kubernetes 1.14 if you're using the latest version of Helm. I'm not sure if there's a mitigation that we can provide at the Helm level, but if someone wants to try tackling this at the client-go level, that would be lovely. That way we can see if there's a resolution that could fix this for all clients going forward.

**Comment by dploeger** [comment]
Yes, I can confirm that helm with kubectl 1.14.2 works like a charm. No requests for tokens anymore. Thanks. I think, this should be a docs issue then, right?

**Comment by sylr** [comment]
I've tested with kubectl 1.14.2 and helm 2.13.0 and I can still reproduce:

```
sylvain@ubuntu-1604-dev:~/src$ helm-v2.13.0 --kube-context=k8s11-euw-dev --tiller-namespace=api-management list
NAME                    REVISION        UPDATED                         STATUS          CHART                   APP VERSION     NAMESPACE
...
...
sylvain@ubuntu-1604-dev:~/src$ kubectl --context=k8s11-euw-dev -n api-management get pods
To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code FV2GA542X to authenticate.
NAME                                 READY   STATUS    RESTARTS   AGE
...
...
sylvain@ubuntu-1604-dev:~/src$ helm-v2.13.0 --kube-context=k8s11-euw-dev --tiller-namespace=api-management list
NAME                    REVISION        UPDATED                         STATUS          CHART                   APP VERSION     NAMESPACE
...
...
sylvain@ubuntu-1604-dev:~/src$ kubectl --context=k8s11-euw-dev -n api-management get pods
To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code D2SC6RNGN to authenticate.
```

**Comment by bacongobbler** [comment]
If you're using kubectl 1.14, please use helm 2.14 as that is compiled against client-go with Kubernetes 1.14 support. You're likely bumping into the same problem; 1.13 and 1.14 are incompatible with eachother.

**Comment by carlossg** [comment]
crossposting from https://github.com/Azure/AKS/issues/1008

If you are stuck with an older helm version, my workaround is to create a new shell session for helm, copy my `.kube/config` as `.kube/config-helm` and set `KUBECONFIG=.kube/config-helm`
Then I can use helm in that session and `kubectl` in other sessions with no issues

But yes, it is solved by upgrading helm

**Comment by hickeyma** [comment]
@dploeger As this is fixed in Helm 2.14, I think the content in this issue should be adequate for users (if they encounter the issue) as they are more than likely to upgrade to the latest versions of Helm. Helm is currently at 2.14.3 and 2.15 will be released shortly.

Closing as fixed in 2.14.

**Comment by carlossg** [comment]
For the record, this is probably caused by #5665
